File: core/network/src/main/java/xyz/aeroitems/core/network/redis/RedisManager.java
package xyz.aeroitems.core.network.redis;

import io.lettuce.core.RedisClient;
import io.lettuce.core.api.StatefulRedisConnection;
import io.lettuce.core.api.async.RedisAsyncCommands;
import io.lettuce.core.codec.ByteArrayCodec;
import io.lettuce.core.codec.RedisCodec;
import io.lettuce.core.codec.StringCodec;

public final class RedisManager {

    private final RedisClient client;
    private final StatefulRedisConnection<String, byte[]> connection;

    public RedisManager(String uri) {
        this.client = RedisClient.create(uri);
        this.connection = client.connect(
                RedisCodec.of(StringCodec.UTF8, ByteArrayCodec.INSTANCE)
        );
    }

    public RedisAsyncCommands<String, byte[]> async() {
        return connection.async();
    }

    public void shutdown() {
        connection.close();
        client.shutdown();
    }
}

File: proxy/velocity/src/main/java/xyz/aeroitems/proxy/velocity/network/RedisManager.kt
package xyz.aeroitems.proxy.velocity.network

import io.lettuce.core.RedisClient
import io.lettuce.core.pubsub.StatefulRedisPubSubConnection
import io.lettuce.core.pubsub.api.sync.RedisPubSubCommands
import io.lettuce.core.pubsub.RedisPubSubAdapter

class RedisManager(uri: String) {

    private val client: RedisClient = RedisClient.create(uri)

    private val pubSubConnection: StatefulRedisPubSubConnection<String, String> =
        client.connectPubSub()

    private val pubSubCommands: RedisPubSubCommands<String, String> =
        pubSubConnection.sync()

    fun addListener(handler: (String) -> Unit) {
        pubSubConnection.addListener(object : RedisPubSubAdapter<String, String>() {
            override fun message(channel: String, message: String) {
                handler(message)
            }
        })
    }

    fun subscribe(channel: String) {
        pubSubCommands.subscribe(channel)
    }

    fun publish(channel: String, message: String) {
        client.connect().use {
            it.sync().publish(channel, message)
        }
    }

    fun shutdown() {
        pubSubConnection.close()
        client.shutdown()
    }
}
